bmq-sdk/src/main/java/com/bloomberg/bmq/impl/infr/msg/StreamParameters.java

```diff
@@ -0,0 +1,142 @@
+import com.bloomberg.bmq.SubscriptionHandle;
+import java.util.Map;
+
+public class StreamParameters {
+
+    public static final int CONSUMER_PRIORITY_INVALID = Integer.MIN_VALUE;
+    public static final String DEFAULT_APP_ID = "__default";
+
+    private String appId;
+    private Subscription[] subscriptions;
+
+    public StreamParameters() {
+        init();
+    }
+
+    public void reset() {
+        init();
+    }
+
+    private void init() {
+        appId = DEFAULT_APP_ID;
+        subscriptions = new Subscription[] {};
+    }
+
+    public String appId() {
+        return appId;
+    }
+
+    public void setAppId(String value) {
+        appId = value;
+    }
+
+    public Subscription[] subscriptions() {
+        return subscriptions;
+    }
+
+    public void setSubscriptions(Subscription[] value) {
+        subscriptions = value;
+    }
+
+    public static StreamParameters createDeconfigureParameters(SubQueueIdInfo subQueueIdInfo) {
+        StreamParameters params = new StreamParameters();
+
+        if (subQueueIdInfo != null) {
+            params.setAppId(subQueueIdInfo.appId());
+        }
+
+        // Empty subscriptions for deconfigure
+
+        return params;
+    }
+
+    public static StreamParameters createParameters(
+            SubQueueIdInfo subQueueIdInfo, QueueOptions ops) {
+        int consumerPriorityCount = 1;
+
+        StreamParameters params = new StreamParameters();
+
+        if (subQueueIdInfo != null) {
+            params.setAppId(subQueueIdInfo.appId());
+        }
+
+        Subscription[] subscriptions;
+
+        if (ops.getSubscriptions().size() == 0) {
+            Subscription sn = new Subscription();
+
+            ConsumerInfo consumer = new ConsumerInfo();
+            consumer.setConsumerPriority(ops.getConsumerPriority());
+            consumer.setConsumerPriorityCount(consumerPriorityCount);
+            consumer.setMaxUnconfirmedMessages(ops.getMaxUnconfirmedMessages());
+            consumer.setMaxUnconfirmedBytes(ops.getMaxUnconfirmedBytes());
+            sn.setConsumers(new ConsumerInfo[] {consumer});
+
+            Expression expr = new Expression();
+            expr.setText(SubscriptionExpression.k_EXPRESSION_DEFAULT);
+            expr.setVersion(SubscriptionExpression.k_VERSION_DEFAULT.toString().toUpperCase());
+            sn.setExpression(expr);
+
+            subscriptions = new Subscription[] {sn};
+        } else {
+            subscriptions = new Subscription[ops.getSubscriptions().size()];
+            int index = 0;
+
+            for (Map.Entry<Integer, Map.Entry<SubscriptionHandle, com.bloomberg.bmq.Subscription>>
+                    entry : ops.getSubscriptions().entrySet()) {
+                com.bloomberg.bmq.Subscription snConfig = entry.getValue().getValue();
+
+                Subscription sn = new Subscription();
+
+                sn.setOrigin(snConfig);
+                sn.setHandle(entry.getValue().getKey());
+                sn.setId(entry.getKey());
```

Comment:
We need to create mapping between `QueueOptions->Subscription` and protocol `Subscription`.
Protocol `Subscription` doesn't exist at the time when we already have `QueueOptions->Subscription` which is also immutable.
Also the same `QueueOptions` could be reused multiple time to build different subscriptions, so its relation to protocol `Subscription` is MANY <-> ONE.
The only place where we can make this mapping is somewhere near the protocol `Subscription` build, right  here. Either we use transient (non serializable) fields for protocol `Subscription`, or we need to make some difficult changes to this method's return values to return this mapping as a separate thing to the external code.